### PR TITLE
fix https://issues.civicrm.org/jira/browse/CRM-19149

### PR DESCRIPTION
**Pull request by wahyukodar**
Issue on JIRA https://issues.civicrm.org/jira/browse/CRM-19149

### CODE REVIEW

**Comment by totten** [comment]
jenkins, ok to test

**Comment by JoeMurray** [comment]
@wahyukodar please fix failing tests
@pradpnayak please QA

**Comment by seamuslee001** [comment]
@wahyukodar Wahyu can you please do git rebase -i <core repo alias>/master and only pick 	d84d83f 6861dfe

**Comment by wahyukodar** [comment]
@seamuslee001 Hi, can you help me what command to write on the terminal?

**Comment by seamuslee001** [comment]
@wahyukodar Wahyu

What you want to do is in your terminal cd to the civicrm root. 
Then do git fetch alias for civicrm core repo if you have it otherwise add another origin civicrm pointing to civicrm core

then do git rebase -i civirepo-alias/master and then follow the prompts.  as per the docs here https://git-scm.com/docs/git-rebase

**Comment by wahyukodar** [comment]
@seamuslee001 Hi, I'm done git rebase, but it's still fail test.. Alert in jenkins "No report files were found. Configuration error".. What should I do?

**Comment by seamuslee001** [comment]
@wahyukodar can you do another git rebase but when you get the list of commits ONLY include the ones relevant to this Pull and no merge commits.

**Comment by wahyukodar** [comment]
@seamuslee001 , when I run git rebase -i civirepo/master, I get 

pick 122250e fix issue https://issues.civicrm.org/jira/browse/CRM-19149
pick 34f6cee CRM-19275 - DB Error on adding priceset when deferred revenue setting is enabled
pick b97c9c4 English grammar edit
pick b86587c CRM-19290 Fix date creation in Karma test
pick c151a36 fix issue https://issues.civicrm.org/jira/browse/CRM-19149
pick 0aa19c5 CRM-19290 Fix date creation in Karma test
pick 8d88ae4 fix https://issues.civicrm.org/jira/browse/CRM-19149
pick d84d83f fix https://issues.civicrm.org/jira/browse/CRM-19149
pick 6861dfe fix https://issues.civicrm.org/jira/browse/CRM-19149
pick 5db0520 fixed unit test
pick c67d5d7 fixed unit test

**Comment by seamuslee001** [comment]
ok so what you want to do is to do

pick 122250e fix issue https://issues.civicrm.org/jira/browse/CRM-19149
#pick 34f6cee CRM-19275 - DB Error on adding priceset when deferred revenue setting is enabled
#pick b97c9c4 English grammar edit
#pick b86587c CRM-19290 Fix date creation in Karma test
pick c151a36 fix issue https://issues.civicrm.org/jira/browse/CRM-19149
#pick 0aa19c5 CRM-19290 Fix date creation in Karma test
pick 8d88ae4 fix https://issues.civicrm.org/jira/browse/CRM-19149
pick d84d83f fix https://issues.civicrm.org/jira/browse/CRM-19149
pick 6861dfe fix https://issues.civicrm.org/jira/browse/CRM-19149
pick 5db0520 fixed unit test
pick c67d5d7 fixed unit test

**Comment by seamuslee001** [comment]
so add in the # where i have suggested and then do git push origin master -f

**Comment by wahyukodar** [comment]
I can't push, this log my terminal

> Wahyu-MacBook-Air:civicrm-core wahyu$ git push origin master -f
> Everything up-to-date
> Wahyu-MacBook-Air:civicrm-core wahyu$ git status
> interactive rebase in progress; onto 87881da
> Last commands done (4 commands done):
>    pick 8d88ae4 fix https://issues.civicrm.org/jira/browse/CRM-19149
>    pick d84d83f fix https://issues.civicrm.org/jira/browse/CRM-19149
>   (see more in file .git/rebase-merge/done)
> Next commands to do (3 remaining commands):
>    pick 6861dfe fix https://issues.civicrm.org/jira/browse/CRM-19149
>    pick 5db0520 fixed unit test
>   (use "git rebase --edit-todo" to view and edit)
> You are currently rebasing branch 'master' on '87881da'.
>   (all conflicts fixed: run "git rebase --continue")
> 
> nothing to commit, working directory clean

**Comment by seamuslee001** [comment]
ok do git commit --allow-empty and then git rebase --continue

**Comment by wahyukodar** [comment]
git rebase --continue error

> git commit --allow-empty
> [detached HEAD e723092] fix https://issues.civicrm.org/jira/browse/CRM-19149
>  Date: Sun Sep 4 17:47:21 2016 +0700
> Wahyu-MacBook-Air:civicrm-core wahyu$ git rebase --continue
> error: could not apply 6861dfe... fix https://issues.civicrm.org/jira/browse/CRM-19149
> 
> When you have resolved this problem, run "git rebase --continue".
> If you prefer to skip this patch, run "git rebase --skip" instead.
> To check out the original branch and stop rebasing, run "git rebase --abort".
> Could not apply 6861dfed5773e9abdb0962a49f883768e8f7a1d8... fix https://issues.civicrm.org/jira/browse/CRM-19149

**Comment by seamuslee001** [comment]
ok now do git mergetool and you should see some errors like <<< etc and remove and fix merge and then do git rebase --continue

**Comment by seamuslee001** [comment]
nice work @wahyukodar :-)

**Comment by wahyukodar** [comment]
@seamuslee001  Thank you very much.. :)

**Comment by tamarmeir** [comment]
Tested, looks good, see screen shot of accounting batch:

negative_contribution_edit_1

Note, however, that the financial transaction still shown with the contribution is the original one for check even after updating the payment method to credit card - not sure why I'm not seeing the other two financial transactions
negative_contribution_edit_2

Looking at civicrm_financial_trx via api explorer, I also noticed that the positive transaction (id = 94) with payment instrument id = 4 is still showing a negative net amount - not sure what that would impact but I'm thinking for consistency's sake, it would be good if the net amount was updated as well:

{

	"is_error":0,
	"version":3,
	"count":3,
	"values":[{
		"id":"94",
		"to_financial_account_id":"6",
		"trxn_date":"2016-09-08 17:18:00",
		"total_amount":"-100.00",
		"fee_amount":"0.00",
		"net_amount":"-100.00",
		"currency":"USD",
		"is_payment":"1",
		"status_id":"1",
		"payment_instrument_id":"4"
	},
	{
		"id":"95",
		"to_financial_account_id":"12",
		"trxn_date":"2016-09-08 17:18:00",
		"total_amount":"100.00",
		"fee_amount":"0.00",
		"net_amount":"-100.00",
		"currency":"USD",
		"is_payment":"1",
		"status_id":"1",
		"payment_instrument_id":"4"
	},
	{
		"id":"96",
		"to_financial_account_id":"12",
		"trxn_date":"2016-09-08 17:18:00",
		"total_amount":"-100.00",
		"fee_amount":"0.00",
		"net_amount":"-100.00",
		"currency":"USD",
		"is_payment":"1",
		"status_id":"1",
		"payment_instrument_id":"1"
	}]
}

Thanks in advance for your time in working on this and happy to test once the above is remedied or if you feel that a separate ticket should be created, please let me now and I will do so.

Thanks!

**Comment by seamuslee001** [comment]
@tamarmeir with the latest commits i think your problems are fixed

**Comment by tamarmeir** [comment]
Almost there - financial transactions associated with contribution look good:

negative_contribution_edit_1a

But api explore is still showing issues with the net amount of the financial transactions created upon editing - this time, it's the negative credit card transaction (id = 96) that shows a positive amount in the "net amount" field:

	"is_error":0,
	"version":3,
	"count":3,
	"values":[{
		"id":"94",
		"to_financial_account_id":"6",
		"trxn_date":"2016-09-14 23:02:00",
		"total_amount":"-100.00",
		"fee_amount":"0.00",
		"net_amount":"-100.00",
		"currency":"USD",
		"is_payment":"1",
		"status_id":"1",
		"payment_instrument_id":"4"
	},
	{
		"id":"95",
		"to_financial_account_id":"6",
		"trxn_date":"2016-09-14 23:02:00",
		"total_amount":"100.00",
		"fee_amount":"0.00",
		"net_amount":"100.00",
		"currency":"USD",
		"is_payment":"1",
		"status_id":"1",
		"payment_instrument_id":"4"
	},
	{
		"id":"96",
		"to_financial_account_id":"12",
		"trxn_date":"2016-09-14 23:02:00",
		"total_amount":"-100.00",
		"fee_amount":"0.00",
		"net_amount":"100.00",
		"currency":"USD",
		"is_payment":"1",
		"status_id":"1",
		"payment_instrument_id":"1"
	}]
}

File: CRM/Contribute/BAO/Contribution.php [hunk 1]
```diff
@@ -3462,21 +3462,34 @@ public static function updateFinancialAccounts(&$params, $context = NULL, $skipT
...
+            $params['trxnParams']['to_financial_account_id'] = $params['to_financial_account_id'];
+            $params['trxnParams']['payment_instrument_id'] = $params['contribution']->payment_instrument_id;
+          }
         }
       }
       else {
-        $params['trxnParams']['to_financial_account_id'] = $params['to_financial_account_id'];
-        $params['trxnParams']['payment_instrument_id'] = $params['contribution']->payment_instrument_id;
+        if ($params['total_amount'] < 0) {
+          $params['trxnParams']['net_amount'] = $params['trxnParams']['total_amount'];
```
Comment (seamuslee001) [inline CRM/Contribute/BAO/Contribution.php]: @wahyukodar Wahyu could you move this line up to L3645 i.e. just after the elseif, That should then solve the last issue found by Tamar

Comment (wahyukodar) [inline CRM/Contribute/BAO/Contribution.php]: Line 3645? are you sure?

Comment (seamuslee001) [inline CRM/Contribute/BAO/Contribution.php]: @wahyukodar good pickup i mean L3465 :-) *reads closer*

Comment (wahyukodar) [inline CRM/Contribute/BAO/Contribution.php]: Ok, I will move thi

>  $params['trxnParams']['net_amount'] = $params['trxnParams']['total_amount'];
>   @seamuslee001

to L3465? Right?

**Comment by tamarmeir** [comment]
Houston, we have lift off with the financial transactions! Not sure why there is a result of "All checks failed" in the last merge build - I will leave it to you, though since you have effectively resolved this issue with the last changes made.

Thanks and kind regards,
Tamar

**Comment by seamuslee001** [comment]
@totten @monishdeb I think this can be merged now tested pretty thoroughly

**Comment by yashodha** [comment]
Jenkins test this please

**Comment by yashodha** [comment]
unrelated failure

File: CRM/Contribute/BAO/Contribution.php [hunk 2]
```diff
@@ -3592,6 +3604,28 @@ public static function updateFinancialAccounts(&$params, $context = NULL, $skipT
         }
       }
     }
+    if ($context == 'changePaymentInstrument') {
+      foreach ($params['line_item'] as $lineitems) {
+        foreach ($lineitems as $fieldValueId => $fieldValues) {
+          $prevFinancialItem = CRM_Financial_BAO_FinancialItem::getPreviousFinancialItem($fieldValues['id']);
+          if (!CRM_Utils_Rule::currencyCode($trxn->currency)) {
+            $trxn->currency = CRM_Core_Config::singleton()->defaultCurrency;
+          }
```
Comment (pradpnayak) [inline CRM/Contribute/BAO/Contribution.php]: I think line 3611-3613 is not required.

File: CRM/Contribute/BAO/Contribution.php [hunk 3]
```diff
@@ -3592,6 +3604,28 @@ public static function updateFinancialAccounts(&$params, $context = NULL, $skipT
...
+
+          // save to entity_financial_trxn table
+          $entityFinancialTrxnParams = array(
+            'entity_table' => "civicrm_financial_item",
+            'entity_id' => $prevFinancialItem->id,
+            'financial_trxn_id' => $trxn->id,
+            'amount' => $trxn->total_amount,
+            'currency' => $trxn->currency,
```
Comment (pradpnayak) [inline CRM/Contribute/BAO/Contribution.php]: civicrm_entity_financial_trxn doesn't have currency column.

File: CRM/Contribute/BAO/Contribution.php [hunk 4]
```diff
@@ -3592,6 +3604,28 @@ public static function updateFinancialAccounts(&$params, $context = NULL, $skipT
...
+          );
+          $entityTrxn = new CRM_Financial_DAO_EntityFinancialTrxn();
+          $entityTrxn->copyValues($entityFinancialTrxnParams);
+          $entityTrxn->save();
```
Comment (pradpnayak) [inline CRM/Contribute/BAO/Contribution.php]: Why are we not using api here?

File: CRM/Contribute/BAO/Contribution.php [hunk 5]
```diff
@@ -3592,6 +3604,28 @@ public static function updateFinancialAccounts(&$params, $context = NULL, $skipT
         }
       }
     }
+    if ($context == 'changePaymentInstrument') {
```
Comment (pradpnayak) [inline CRM/Contribute/BAO/Contribution.php]: Is this as per the dataflow for accounting? 
https://wiki.civicrm.org/confluence/display/CRM/CiviAccounts+Data+Flow#CiviAccountsDataFlow-ChangePaymentInstrumentforaPartlyPaidorCompletedcontribution

Comment (seamuslee001) [inline CRM/Contribute/BAO/Contribution.php]: @pradpnayak all this is doing is adding in a linkage back to the financial_item without it the display on the contribution tab does not show all financial_trxns

File: CRM/Contribute/BAO/Contribution.php (hunk 4 above)
Comment (seamuslee001) [inline CRM/Contribute/BAO/Contribution.php]: probably better idea

File: CRM/Contribute/BAO/Contribution.php (hunk 3 above)
Comment (seamuslee001) [inline CRM/Contribute/BAO/Contribution.php]: this is as per https://github.com/civicrm/civicrm-core/blob/2fa84df330aecce758529af45c2e1986b025e178/CRM/Core/BAO/FinancialTrxn.php#L69

File: CRM/Contribute/BAO/Contribution.php (hunk 5 above)
Comment (pradpnayak) [inline CRM/Contribute/BAO/Contribution.php]: yeah, agreed. Can you please update the wiki and also inform @JoeMurray about the changes?

File: CRM/Contribute/BAO/Contribution.php (hunk 3 above)
Comment (pradpnayak) [inline CRM/Contribute/BAO/Contribution.php]: yeah, its not required in either of the place. Can you remove them as cruft?

File: CRM/Contribute/BAO/Contribution.php (hunk 5 above)
Comment (seamuslee001) [inline CRM/Contribute/BAO/Contribution.php]: sure will give it a shot @pradpnayak

File: CRM/Contribute/BAO/Contribution.php (hunk 3 above)
Comment (seamuslee001) [inline CRM/Contribute/BAO/Contribution.php]: sure I'll do a tidy up PR

File: CRM/Contribute/BAO/Contribution.php [hunk 6]
```diff
@@ -3462,21 +3462,34 @@ public static function updateFinancialAccounts(&$params, $context = NULL, $skipT
...
+        if ($params['total_amount'] < 0) {
+          $params['trxnParams']['payment_instrument_id'] = $params['prevContribution']->payment_instrument_id;
+          $params['trxnParams']['to_financial_account_id'] = CRM_Core_DAO::getFieldValue('CRM_Financial_DAO_FinancialTrxn', $lastFinancialTrxnId['financialTrxnId'], 'to_financial_account_id');
+        }
+        else {
+          $params['trxnParams']['to_financial_account_id'] = $params['to_financial_account_id'];
+          $params['trxnParams']['payment_instrument_id'] = $params['contribution']->payment_instrument_id;
+        }
       }
     }
```
Comment (pradpnayak) [inline CRM/Contribute/BAO/Contribution.php]: @seamuslee001 line 3474-3481 and 3484-3492 looks duplicate. Can we re-write line from 3472-3492 as
`  if (!empty($lastFinancialTrxnId['financialTrxnId'])) {
        if ($params['total_amount'] != $params['trxnParams']['total_amount']) {
          $params['trxnParams']['to_financial_account_id'] = CRM_Core_DAO::getFieldValue('CRM_Financial_DAO_FinancialTrxn', $lastFinancialTrxnId['financialTrxnId'], 'to_financial_account_id');
          $params['trxnParams']['payment_instrument_id'] = $params['prevContribution']->payment_instrument_id;
        }
        else {
          $params['trxnParams']['to_financial_account_id'] = $params['to_financial_account_id'];
          $params['trxnParams']['payment_instrument_id'] = $params['contribution']->payment_instrument_id;
        }
      }
    }`

Comment (seamuslee001) [inline CRM/Contribute/BAO/Contribution.php]: @pradpnayak see followup pr https://github.com/civicrm/civicrm-core/pull/9055

**Pull request merged**